### PR TITLE
Add capability of execute known trajectory with a ROS action

### PR DESCRIPTION
**Pull request by wkentaro**
Proposed at https://github.com/ros-planning/moveit_ros/pull/709#discussion_r70888457 https://github.com/ros-planning/moveit_ros/pull/713#issuecomment-232848243
Depends on https://github.com/ros-planning/moveit_msgs/pull/23

### CODE REVIEW

**Comment by davetcoleman** [comment]
The background for why we need this: "In order to allow stopping of a trajectory, the service should be turned into an action as well."

File: move_group/src/default_capabilities/execute_action_capability.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2016, Willow Garage, Inc.
```
Comment (davetcoleman) [inline move_group/src/default_capabilities/execute_action_capability.cpp]: i think you could change the company name to your university or something. WG is not writing code in 2016 ;-)

File: move_group/include/moveit/move_group/capability_names.h [hunk 2]
```diff
@@ -46,6 +46,7 @@ static const std::string PLANNER_SERVICE_NAME = "plan_kinematic_path";    // nam
 static const std::string EXECUTE_SERVICE_NAME = "execute_kinematic_path"; // name of the advertised service (within the ~ namespace)
 static const std::string QUERY_PLANNERS_SERVICE_NAME = "query_planner_interface"; // name of the advertised query planners service
 static const std::string MOVE_ACTION = "move_group"; // name of 'move' action
+static const std::string EXECUTE_ACTION = "execute_action"; // name of 'execute' action
```
Comment (davetcoleman) [inline move_group/include/moveit/move_group/capability_names.h]: i think "execute_path" or "execute_traj" would be better name. The other actions are not post-fixed with action

File: move_group/src/default_capabilities/execute_action_capability.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Kentaro Wada */
```
Comment (davetcoleman) [inline move_group/src/default_capabilities/execute_action_capability.h]: short description of why you've added this/what this does?

File: planning_interface/move_group_interface/src/move_group.cpp [hunk 4]
```diff
@@ -48,6 +48,7 @@
 #include <moveit/common_planning_interface_objects/common_objects.h>
 #include <moveit/robot_state/conversions.h>
 #include <moveit_msgs/MoveGroupAction.h>
+#include <moveit_msgs/ExecuteKnownTrajectoryAction.h>
```
Comment (davetcoleman) [inline planning_interface/move_group_interface/src/move_group.cpp]: seems like there is a lot of variance in your naming scheme:
- ExecuteKnownTrajectoryAction.mg
- MoveGroupExecuteAction
- "ExecutePathAction"
- ExecuteActionCapability
- MoveGroupExecuteActionCapabaility

Let's unify the naming under a main theme ("ExecutePath"?) and name everything based off that

File: planning_interface/move_group_interface/src/move_group.cpp [hunk 5]
```diff
@@ -145,7 +151,7 @@ class MoveGroup::MoveGroupImpl
                                                                                            false));
     waitForAction(place_action_client_, wait_for_server, move_group::PLACE_ACTION);
 
-    execute_service_ = node_handle_.serviceClient<moveit_msgs::ExecuteKnownTrajectory>(move_group::EXECUTE_SERVICE_NAME);
+    // execute_service_ = node_handle_.serviceClient<moveit_msgs::ExecuteKnownTrajectory>(move_group::EXECUTE_SERVICE_NAME);
```
Comment (davetcoleman) [inline planning_interface/move_group_interface/src/move_group.cpp:150]: Is this backwards compatibile?

**Comment by rhaschke** [comment]
@wkentaro I started to work along that direction too. In my [comment](https://github.com/ros-planning/moveit_ros/pull/717#issue-165934609) for #717 I wrote:

> An alternative, IMHO even better approach is to turn the service into an action, which provides feedback and can be explicitly interrupted. I did that, but it requires several changes to existing moveit_config packages: as the service capability would be renamed (from MoveGroupExecuteService to MoveGroupExecuteAction all move_group.launch files would need an update.

I pushed two new branches:
- [traj-exec-action](/ros-planning/moveit_ros/tree/traj-exec-action): replace the service by an action (doesn't work without major changes as indicated)
- [new-traj-exec-action](/ros-planning/moveit_ros/tree/new-traj-exec-action): add the action additionally to the service

What is missing is the following:

> However, I could think of a slow transition, adding the new capability in Kinetic and loading both the old and new capability in newly created move_config packages (needs an adaption in setup assistant). In some later release we could remove the old service capability, hoping that meanwhile everybody has adapted their move configs.

If you could work along this direction, that would be indeed great. Consider merging your implementation and mine from [new-traj-exec-action](/ros-planning/moveit_ros/tree/new-traj-exec-action). I provide feedback about the finished trajectory segment.

File: planning_interface/move_group_interface/src/move_group.cpp (hunk 4 above)
Comment (wkentaro) [inline planning_interface/move_group_interface/src/move_group.cpp]: Sounds nice. The class name should be `MoveGroupExecutePathAction`?

Comment (wkentaro) [inline planning_interface/move_group_interface/src/move_group.cpp]: Fixed.

File: move_group/src/default_capabilities/execute_action_capability.h (hunk 3 above)
Comment (wkentaro) [inline move_group/src/default_capabilities/execute_action_capability.h]: Fixed.

File: move_group/include/moveit/move_group/capability_names.h (hunk 2 above)
Comment (wkentaro) [inline move_group/include/moveit/move_group/capability_names.h]: Fixed.

File: move_group/src/default_capabilities/execute_action_capability.cpp (hunk 1 above)
Comment (wkentaro) [inline move_group/src/default_capabilities/execute_action_capability.cpp]: Fixed.

File: planning_interface/move_group_interface/src/move_group.cpp (hunk 5 above)
Comment (wkentaro) [inline planning_interface/move_group_interface/src/move_group.cpp:150]: The problem is the default capabilities are given by rosparam at here.
 https://github.com/ros-planning/moveit_ros/blob/kinetic-devel/move_group/src/move_group.cpp#L107

I was not sure how the execute service could be replaced by the action.
But in @rhaschke 's [comment](https://github.com/ros-planning/moveit_ros/pull/719#issuecomment-233404206), he suggests the process of migration and I will follow that.

**Comment by wkentaro** [comment]
> Consider merging your implementation and mine from new-traj-exec-action. I provide feedback about the finished trajectory segment.

Ok, I will do this tomorrow.

**Comment by v4hn** [comment]
What's the state of this request?

**Comment by rhaschke** [comment]
I handed over to @wkentaro. Did you have time to work on it?

**Comment by wkentaro** [comment]
@v4hn @rhaschke Sorry for late reply. I tried to merge the implementation in new-traje-action branch, and I think I've done. Could you please review?

**Comment by v4hn** [comment]
@wkentaro Could you port this request to https://github.com/ros-planning/moveit ?
We are not going to merge anything into this repository anymore.

**Comment by wkentaro** [comment]
Oh, moveit is being consolidated! Alright, I will do it.

**Comment by v4hn** [comment]
On Tue, Aug 09, 2016 at 11:49:04PM -0700, Kentaro Wada wrote:

> Oh, moveit is being consolidated!

_has been_ :)

**Comment by wkentaro** [comment]
Ported this PR to https://github.com/ros-planning/moveit/pull/60.

**Pull request closed without merging**